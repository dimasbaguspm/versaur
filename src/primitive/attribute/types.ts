import type { ReactNode, HTMLAttributes } from 'react'

/**
 * Props for the Attribute component
 */
export interface AttributeProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * The attribute title (displayed as h4)
   */
  title: string
  /**
   * The attribute content (displayed as p)
   */
  children: ReactNode

  /**
   * Whether the attribute has bottom margin
   * @default false
   */
  hasMargin?: boolean
}
