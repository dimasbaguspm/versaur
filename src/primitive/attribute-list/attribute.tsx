import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import type { AttributeListProps } from './types'
import { AttributeListItem } from './attribute.atoms'
import { getGridCols } from './helpers'

/**
 * Compound AttributeList component for Versaur UI
 *
 * Usage:
 * <AttributeList columns={4}>
 *   <AttributeList.Item span={4} title='Full Width'>Full width content</AttributeList.Item>
 *   <AttributeList.Item span={2} title='Half Width'>Half width content</AttributeList.Item>
 *   <AttributeList.Item span={2} title='Half Width'>Half width content</AttributeList.Item>
 * </AttributeList>
 */
const AttributeListRoot = forwardRef<HTMLUListElement, AttributeListProps>(
  function AttributeListRoot(
    { children, className, columns = 4, ...props },
    ref
  ) {
    const gridCols = getGridCols(columns)

    return (
      <ul
        ref={ref}
        className={cn('grid gap-4', gridCols, className)}
        {...props}
      >
        {children}
      </ul>
    )
  }
)

export const AttributeList = Object.assign(AttributeListRoot, {
  Item: AttributeListItem,
})
