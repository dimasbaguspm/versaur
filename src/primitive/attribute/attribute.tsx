import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import type { AttributeProps } from './types'
import { Text } from '../text'

/**
 * Attribute component for Versaur UI
 *
 * A simple component that displays a title and content in a structured format.
 *
 * Usage:
 * <Attribute title="Name">John Doe</Attribute>
 */
export const Attribute = forwardRef<HTMLDivElement, AttributeProps>(
  function Attribute({ title, children, className, hasMargin, ...props }, ref) {
    return (
      <div
        ref={ref}
        className={cn('space-y-1', className, hasMargin && 'mb-4')}
        {...props}
      >
        <Text
          as='h4'
          fontSize='xs'
          fontWeight='normal'
          color='gray'
          className='leading-none'
        >
          {title}
        </Text>
        <Text as='p' fontSize='sm' fontWeight='normal'>
          {children}
        </Text>
      </div>
    )
  }
)
