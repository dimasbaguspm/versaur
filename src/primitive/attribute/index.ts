export { Attribute } from './attribute'
export type { AttributeProps } from './types'
