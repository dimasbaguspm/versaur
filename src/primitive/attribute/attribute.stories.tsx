import type { Meta, StoryObj } from '@storybook/react'
import { Attribute } from './attribute'

const meta: Meta<typeof Attribute> = {
  title: 'Primitive/Attribute',
  component: Attribute,
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component:
          'Attribute components for displaying structured information. The Attribute component displays a title-content pair, while AttributeList provides a grid-based layout for multiple attributes.',
      },
    },
  },
  tags: ['autodocs'],
  argTypes: {
    title: {
      control: 'text',
      description: 'The attribute title displayed as an h4 element',
    },
    children: {
      control: 'text',
      description: 'The attribute content displayed as a p element',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

/**
 * Basic usage of the Attribute component with a simple title and content.
 */
export const Default: Story = {
  args: {
    title: 'Name',
    children: 'John Doe',
  },
}

/**
 * Attribute with longer content to demonstrate text wrapping.
 */
export const LongContent: Story = {
  args: {
    title: 'Description',
    children:
      'This is a longer description that demonstrates how the attribute component handles text wrapping and multiple lines of content.',
  },
}
