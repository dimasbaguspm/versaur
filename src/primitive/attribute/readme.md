# Attribute

The `Attribute` component is used to display structured information with a title-content pair

## Basic Usage

### Attribute Component

```tsx
import { Attribute } from '@versaur/primitive/attribute'

export default function UserProfile() {
  return (
    <div className='space-y-4'>
      <Attribute title='Name'>John Doe</Attribute>
      <Attribute title='Email'>john.doe@example.com</Attribute>
      <Attribute title='Location'>New York, NY</Attribute>
    </div>
  )
}
```
