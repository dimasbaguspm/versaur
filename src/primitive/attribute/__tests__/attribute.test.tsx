import { render } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../attribute.stories'

const { Default, LongContent } = composeStories(stories)

describe('Attribute', () => {
  it('should render with correct HTML structure', () => {
    expect(render(<Default />).asFragment()).toMatchSnapshot()
  })

  it('should render with long content', () => {
    const { getByText, getByRole } = render(<LongContent />)
    expect(getByRole('heading', { level: 4 })).toHaveTextContent('Description')
    expect(getByText(/This is a longer description/)).toBeInTheDocument()
  })
})
