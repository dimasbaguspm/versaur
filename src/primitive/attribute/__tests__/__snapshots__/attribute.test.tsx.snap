// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Attribute > should render with correct HTML structure 1`] = `
<DocumentFragment>
  <div
    class="space-y-1"
  >
    <h4
      class="text-gray-500 text-left text-sm font-normal leading-none"
    >
      Name
    </h4>
    <p
      class="text-ghost text-left text-base font-normal"
    >
      John Doe
    </p>
  </div>
</DocumentFragment>
`;
